server/src/server.rs

```diff
@@ -36,26 +37,161 @@ fn error(error: cubeb::Error) -> ClientMessage {
+            devtype.contains(cubeb::DeviceType::INPUT),
+            self.input_registered,
+        ) {
+            (true, false) => self.internal_register(context, cubeb::DeviceType::INPUT),
+            (false, true) => self.internal_unregister(context, cubeb::DeviceType::INPUT),
+            _ => {}
+        }
+        match (
+            devtype.contains(cubeb::DeviceType::OUTPUT),
+            self.output_registered,
+        ) {
+            (true, false) => self.internal_register(context, cubeb::DeviceType::OUTPUT),
+            (false, true) => self.internal_unregister(context, cubeb::DeviceType::OUTPUT),
+            _ => {}
+        }
+    }
+
+    fn internal_register(&mut self, context: &cubeb::Context, devtype: cubeb::DeviceType) {
+        let user_ptr = self as *const CubebDeviceCollectionManager as *mut c_void;
+        unsafe {
+            if devtype.contains(cubeb::DeviceType::INPUT) {
+                assert_eq!(self.input_registered, false);
+                context
+                    .register_device_collection_changed(
+                        cubeb::DeviceType::INPUT,
+                        Some(device_collection_changed_input_cb_c),
+                        user_ptr,
+                    )
+                    .expect("input devcol register failed");
+                self.input_registered = true;
+            }
+            if devtype.contains(cubeb::DeviceType::OUTPUT) {
+                assert_eq!(self.output_registered, false);
+                context
+                    .register_device_collection_changed(
+                        cubeb::DeviceType::OUTPUT,
+                        Some(device_collection_changed_output_cb_c),
+                        user_ptr,
+                    )
+                    .expect("output devcol register failed");
+                self.output_registered = true;
+            }
+        }
+    }
+
+    fn internal_unregister(&mut self, context: &cubeb::Context, devtype: cubeb::DeviceType) {
+        unsafe {
+            if devtype.contains(cubeb::DeviceType::INPUT) {
+                assert_eq!(self.input_registered, true);
+                context
+                    .register_device_collection_changed(
+                        cubeb::DeviceType::INPUT,
+                        None,
+                        std::ptr::null_mut(),
+                    )
+                    .expect("input devcol unregister failed");
+                self.input_registered = false;
+            }
+            if devtype.contains(cubeb::DeviceType::OUTPUT) {
+                assert_eq!(self.output_registered, true);
+                context
+                    .register_device_collection_changed(
+                        cubeb::DeviceType::OUTPUT,
+                        None,
+                        std::ptr::null_mut(),
+                    )
+                    .expect("output devcol unregister failed");
+                self.output_registered = false;
+            }
+        }
+    }
```

Comment:
Is it possible to merge the duplicated code in `internal_unregister` and `internal_register`?